In [1]:
# This Python 3 environment comes with many helpful analytics libraries installed
# It is defined by the kaggle/python Docker image: https://github.com/kaggle/docker-python
# For example, here's several helpful packages to load

import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)

# Input data files are available in the read-only "../input/" directory
# For example, running this (by clicking run or pressing Shift+Enter) will list all files under the input directory

import os
for dirname, _, filenames in os.walk('/kaggle/input'):
    for filename in filenames:
        print(os.path.join(dirname, filename))

# You can write up to 20GB to the current directory (/kaggle/working/) that gets preserved as output when you create a version using "Save & Run All" 
# You can also write temporary files to /kaggle/temp/, but they won't be saved outside of the current session

In [2]:
!pip install -Uqq fastbook
import fastbook
fastbook.setup_book()

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 719.8/719.8 kB 12.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 124.1/124.1 kB 9.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 193.6/193.6 kB 16.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 246.9/246.9 kB 19.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 363.4/363.4 MB 4.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 664.8/664.8 MB 2.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 211.5/211.5 MB 8.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 56.3/56.3 MB 30.3 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 127.9/127.9 MB 13.3 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 207.5/207.5 MB 8.3 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 21.1/21.1 MB 67.3 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.6/1.6 MB 60.0 MB/s eta 0:00:00
ERROR: pip's dependency resolver does not

In [3]:
from fastai.vision.all import *
from fastbook import *

matplotlib.rc('image', cmap='Greys')

In [4]:
path = untar_data(URLs.MNIST)
Path.BASE_PATH = path
path.ls()

(#2) [Path('testing'),Path('training')]

In [5]:
(path/'training').ls()

(#10) [Path('training/5'),Path('training/6'),Path('training/4'),Path('training/1'),Path('training/9'),Path('training/7'),Path('training/3'),Path('training/2'),Path('training/8'),Path('training/0')]

# Training Set

In [6]:
train_0  = torch.stack([tensor(Image.open(o)) for o in (path/'training'/'0').ls().sorted()]).float()/255
train_1  = torch.stack([tensor(Image.open(o)) for o in (path/'training'/'1').ls().sorted()]).float()/255
train_2  = torch.stack([tensor(Image.open(o)) for o in (path/'training'/'2').ls().sorted()]).float()/255
train_3  = torch.stack([tensor(Image.open(o)) for o in (path/'training'/'3').ls().sorted()]).float()/255
train_4  = torch.stack([tensor(Image.open(o)) for o in (path/'training'/'4').ls().sorted()]).float()/255
train_5  = torch.stack([tensor(Image.open(o)) for o in (path/'training'/'5').ls().sorted()]).float()/255
train_6  = torch.stack([tensor(Image.open(o)) for o in (path/'training'/'6').ls().sorted()]).float()/255
train_7  = torch.stack([tensor(Image.open(o)) for o in (path/'training'/'7').ls().sorted()]).float()/255
train_8  = torch.stack([tensor(Image.open(o)) for o in (path/'training'/'8').ls().sorted()]).float()/255
train_9  = torch.stack([tensor(Image.open(o)) for o in (path/'training'/'9').ls().sorted()]).float()/255

# Validation Set

In [7]:
valid_0 = torch.stack([tensor(Image.open(o)) for o in (path/'testing'/'0').ls()]).float()/255
valid_1 = torch.stack([tensor(Image.open(o)) for o in (path/'testing'/'1').ls()]).float()/255
valid_2 = torch.stack([tensor(Image.open(o)) for o in (path/'testing'/'2').ls()]).float()/255
valid_3 = torch.stack([tensor(Image.open(o)) for o in (path/'testing'/'3').ls()]).float()/255
valid_4 = torch.stack([tensor(Image.open(o)) for o in (path/'testing'/'4').ls()]).float()/255
valid_5 = torch.stack([tensor(Image.open(o)) for o in (path/'testing'/'5').ls()]).float()/255
valid_6 = torch.stack([tensor(Image.open(o)) for o in (path/'testing'/'6').ls()]).float()/255
valid_7 = torch.stack([tensor(Image.open(o)) for o in (path/'testing'/'7').ls()]).float()/255
valid_8 = torch.stack([tensor(Image.open(o)) for o in (path/'testing'/'8').ls()]).float()/255
valid_9 = torch.stack([tensor(Image.open(o)) for o in (path/'testing'/'9').ls()]).float()/255

# Concatinating All Digits

In [8]:
train_x = torch.cat([
    train_0, train_1, train_2, train_3, train_4,
    train_5, train_6, train_7, train_8, train_9
]).unsqueeze(1)

valid_x = torch.cat([
    valid_0, valid_1, valid_2, valid_3, valid_4,
    valid_5, valid_6, valid_7, valid_8, valid_9
]).unsqueeze(1)

In [9]:
train_y = torch.cat([
    torch.full((train_0.shape[0],), 0),
    torch.full((train_1.shape[0],), 1),
    torch.full((train_2.shape[0],), 2),
    torch.full((train_3.shape[0],), 3),
    torch.full((train_4.shape[0],), 4),
    torch.full((train_5.shape[0],), 5),
    torch.full((train_6.shape[0],), 6),
    torch.full((train_7.shape[0],), 7),
    torch.full((train_8.shape[0],), 8),
    torch.full((train_9.shape[0],), 9)
], dim=0).long()

valid_y = torch.cat([
    torch.full((valid_0.shape[0],), 0),
    torch.full((valid_1.shape[0],), 1),
    torch.full((valid_2.shape[0],), 2),
    torch.full((valid_3.shape[0],), 3),
    torch.full((valid_4.shape[0],), 4),
    torch.full((valid_5.shape[0],), 5),
    torch.full((valid_6.shape[0],), 6),
    torch.full((valid_7.shape[0],), 7),
    torch.full((valid_8.shape[0],), 8),
    torch.full((valid_9.shape[0],), 9)
], dim=0).long()

In [10]:
train_dset = list(zip(train_x, train_y))
valid_dset = list(zip(valid_x, valid_y))

In [11]:
train_dl = DataLoader(train_dset, batch_size=256, shuffle=True)
valid_dl = DataLoader(valid_dset, batch_size=256)

# Custom Classes From Scratch

In [12]:
class MyConv2D(nn.Module) :
    def __init__(self, in_ch, out_ch, ker_size, stride=1, padding=0) :
        super().__init__()
        self.in_ch = in_ch
        self.out_ch = out_ch

        if isinstance(ker_size, int) :
            ker_size = (ker_size, ker_size)
        if isinstance(stride, int) :
            stride = (stride, stride)
        if isinstance(padding, int) :
            padding = (padding, padding)

        self.ker_size = ker_size
        self.stride = stride
        self.padding = padding

        # He initialization
        self.weight = nn.Parameter(torch.randn(
          out_ch, in_ch, *ker_size  
        ) * (1.0 / (in_ch * ker_size[0] * ker_size[1])**0.5))

        self.bias = nn.Parameter(torch.zeros(out_ch))

    def forward(self, x) :
        device = x.device
        
        B, C, H, W = x.shape
        kh, kw = self.ker_size

        # Shape: (B, in_ch * kh * kw, L), where L = number of patches
        x_unf = F.unfold(x, kernel_size=self.ker_size, stride=self.stride, padding=self.padding)

        # Reshape weights to (out_ch, in_ch * kh * kw)
        weight_flat = self.weight.view(self.out_ch, -1).to(device)

        out_unf = weight_flat @ x_unf  # shape: (B, out_ch, L)
        out_unf = out_unf + self.bias.view(1, -1, 1).to(device)

        out_h = (H + 2*self.padding[0] - kh) // self.stride[0] + 1
        out_w = (W + 2*self.padding[1] - kw) // self.stride[1] + 1

        out = out_unf.view(B, self.out_ch, out_h, out_w)
        return out

In [13]:
class MyCrossEntropyLoss(nn.Module) :
    def __init__(self) :
        super().__init__()

    def forward(self, logits, targets) :
        # logits: (N, C)
        # targets: (N,)
        log_probs = F.log_softmax(logits, dim=1)
        loss = -log_probs[torch.arange(logits.size(0)), targets]
        return loss.mean()

In [14]:
class MySGD(Optimizer) :
    def __init__(self, params, lr=0.01) :
        self.params = list(params)
        self.lr = lr

    def step(self, closure=None) :
        for p in self.params :
            if p.grad is not None :
                with torch.no_grad() :
                    p -= self.lr * p.grad

    def zero_grad(self) :
        for p in self.params :
            if p.grad is not None :
                p.grad.zero_()

In [15]:
class MyModel(nn.Module) :
    def __init__(self) :
        super().__init__()
        self.conv1 = MyConv2D(1, 4, ker_size=3, padding=1)
        self.bn1   = nn.BatchNorm2d(4)

        self.conv2 = MyConv2D(4, 8, ker_size=3, padding=1)
        self.bn2   = nn.BatchNorm2d(8)

        self.conv3 = MyConv2D(8, 16, ker_size=3, padding=1)
        self.bn3   = nn.BatchNorm2d(16)

        self.flatten = nn.Flatten()
        self.dropout = nn.Dropout(p=0.2)
        self.fc = nn.Linear(16*28*28, 10)

    def forward(self, x) :
        x = F.relu(self.bn1(self.conv1(x)))
        x = F.relu(self.bn2(self.conv2(x)))
        x = F.relu(self.bn3(self.conv3(x)))
        x = self.flatten(x)
        x = self.dropout(x)
        x = self.fc(x)
        return x

In [16]:
device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')

model = MyModel().to(device)
loss_fn = MyCrossEntropyLoss()
optimizer = MySGD(model.parameters(), lr=0.01)

In [17]:
num_epochs = 20

for epoch in range(num_epochs) :
    model.train()
    total_train_loss = 0.0
    total_train_correct = 0
    total_train_samples = 0
    
    for batch_idx, (x_batch, y_batch) in enumerate(train_dl) :
        x_batch = x_batch.to(device)
        y_batch = y_batch.to(device)
        logits = model(x_batch)
        loss = loss_fn(logits, y_batch)
        loss.backward()
        optimizer.step()
        optimizer.zero_grad()

        total_train_loss += loss.item()

        preds = torch.argmax(logits, dim=1)
        total_train_correct = (preds==y_batch).sum().item()
        total_train_samples = y_batch.size(0)

    avg_train_loss = total_train_loss / len(train_dl)
    train_acc = total_train_correct / total_train_samples

    model.eval()
    total_val_loss = 0.0
    total_val_correct = 0
    total_val_samples = 0

    with torch.no_grad() :
        for x_val, y_val in valid_dl :
            x_val = x_val.to(device)
            y_val = y_val.to(device)
            val_logits = model(x_val)
            val_loss = loss_fn(val_logits, y_val)
            total_val_loss += val_loss.item()

            val_preds = torch.argmax(val_logits, dim=1)
            total_val_correct += (val_preds==y_val).sum().item()
            total_val_samples += y_val.size(0)

    avg_val_loss = total_val_loss / len(valid_dl)
    val_acc = total_val_correct / total_val_samples

    print(f"Epoch {epoch+1}")
    print(f"    Train Loss: {avg_train_loss:.4f} | Train Acc: {train_acc:.4f}")
    print(f"    Valid Loss: {avg_val_loss:.4f} | Valid Acc: {val_acc:.4f}")
    print("-" * 50)

Epoch 1
    Train Loss: 0.3573 | Train Acc: 0.9792
    Valid Loss: 0.1879 | Valid Acc: 0.9516
--------------------------------------------------
Epoch 2
    Train Loss: 0.1551 | Train Acc: 0.9583
    Valid Loss: 0.1306 | Valid Acc: 0.9662
--------------------------------------------------
Epoch 3
    Train Loss: 0.1151 | Train Acc: 0.9688
    Valid Loss: 0.1009 | Valid Acc: 0.9733
--------------------------------------------------
Epoch 4
    Train Loss: 0.0936 | Train Acc: 0.9479
    Valid Loss: 0.0868 | Valid Acc: 0.9756
--------------------------------------------------
Epoch 5
    Train Loss: 0.0799 | Train Acc: 0.9896
    Valid Loss: 0.0767 | Valid Acc: 0.9789
--------------------------------------------------
Epoch 6
    Train Loss: 0.0709 | Train Acc: 0.9792
    Valid Loss: 0.0711 | Valid Acc: 0.9794
--------------------------------------------------
Epoch 7
    Train Loss: 0.0637 | Train Acc: 0.9792
    Valid Loss: 0.0680 | Valid Acc: 0.9799
------------------------------------

# Two Layers CNN
Train Loss: 0.0635 | Train Acc: 0.9792 | Valid Loss: 0.0576 | Valid Acc: 0.9831

# Three Layers CNN
Train Loss: 0.0328 | Train Acc: 0.9792 | Valid Loss: 0.0547 | Valid Acc: 0.9836

In [18]:
torch.save(model, 'mnist_model.pkl')